Create a driver to update StageMode Interactions based on Gesture Input

Phipson opened the pull request:
#### ec1c172ffe777a3af45ec795e7bee066fb38b132
<pre>
Create a driver to update StageMode Interactions based on Gesture Input
<a href="https://bugs.webkit.org/show_bug.cgi?id=287820">https://bugs.webkit.org/show_bug.cgi?id=287820</a>
<a href="https://rdar.apple.com/145006262">rdar://145006262</a>

Reviewed by Wenson Hsieh, Richard Robinson, Elliott Williams, and Ada Chan.

This PR adds the StageModeInteractionDriver, which is a helper class object used in the ModelProcessModelPlayerProxy
to control how the model responds to the stagemode transform updates, based on the applied stagemode attribute. When
stagemode is set to &apos;none&apos; we ignore all transforms passed to the player proxy. When stagemode is set to &apos;orbit&apos; we
use the translation from the transform to rotate the model. To separate stagemode updates from Javascript updates,
we introduce an Interaction Container that wraps around the model entity, on which the stagemode updates will be applied.

The driver will be implemented in <a href="https://bugs.webkit.org/show_bug.cgi?id=287601.">https://bugs.webkit.org/show_bug.cgi?id=287601.</a>

* Source/WebKit/ModelProcess/cocoa/ModelProcessModelPlayerProxy.h:
* Source/WebKit/ModelProcess/cocoa/ModelProcessModelPlayerProxy.mm:
(WebKit::ModelProcessModelPlayerProxy::didFinishLoading):
(WebKit::ModelProcessModelPlayerProxy::beginStageModeTransform):
(WebKit::ModelProcessModelPlayerProxy::updateStageModeTransform):
(WebKit::ModelProcessModelPlayerProxy::endStageModeInteraction):
(WebKit::ModelProcessModelPlayerProxy::setStageMode):
(WebKit::ModelProcessModelPlayerProxy::applyStageModeOperationToDriver):
- Updates the callbacks to use the interaction driver
- Also adds helper function to update the driver&apos;s internal state based on the stagemode attribute

* Source/WebKit/UIProcess/Cocoa/WebKitSwiftSoftLink.h:
* Source/WebKit/UIProcess/Cocoa/WebKitSwiftSoftLink.mm:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebKitSwift/StageMode/WKStageMode.swift: Added.
(WKStageModeInteractionDriver.stageModeOperation):
(WKStageModeInteractionDriver.interactionDidBegin(_:)):
(WKStageModeInteractionDriver.interactionDidUpdate(_:)):
(WKStageModeInteractionDriver.interactionDidEnd):
(WKStageModeInteractionDriver.operationDidChange(_:)):
- Adds the necessary APIs to apply the transform to the interaction target.

* Source/WebKit/WebKitSwift/StageMode/WKStageModeBridging.h: Copied from Source/WebKit/WebKitSwift/WebKitSwift.h.
* Source/WebKit/WebKitSwift/WebKitSwift.h:

Canonical link: <a href="https://commits.webkit.org/291028@main">https://commits.webkit.org/291028@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/418af880f9652ee2dd264a4189dd2976ca1db164

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/91664 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/11193 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/733 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/96628 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/42325 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/11568 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/19753 "Built successfully") | [❌ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/70374 "Failure limit exceed. At least found 2 new test failures: imported/w3c/web-platform-tests/html/semantics/the-button-element/interest-target/interesttarget-button-event-dispatch.tentative.html imported/w3c/web-platform-tests/html/semantics/the-button-element/interest-target/interesttarget-on-popover-behavior.tentative.html (failure)") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/27888 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/94665 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/8829 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/83043 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/50700 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/8593 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/41514 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/78909 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/636 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/98637 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/18814 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/13940 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/79401 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/19068 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/78882 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/78612 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/19475 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/23132 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/486 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/11919 "The change is no longer eligible for processing.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/18806 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/24071 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/18514 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/21971 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/20271 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rr-codes reviewed (changes requested):
(marking as Request changes until the WKS part is moved into WK)

achan00 commented on Source/WebKit/ModelProcess/cocoa/ModelProcessModelPlayerProxy.mm:
```diff
@@ -454,17 +455,28 @@ static CGFloat effectivePointsPerMeter(CALayer *caLayer)
+    // This ensures that the transforms applied on the model don't conflict with model-specific animations/JS transforms
+    REPtr interactionTarget = adoptRE(REEntityCreate());
+    REEntitySetName(interactionTarget.get(), "WebKit:InteractionContainerEntity");
+    m_interactionContainerEntity = interactionTarget;
```
Nit: can we initialize `m_interactionContainerEntity` directly without the local variable `interactionTarget`?

achan00 commented on Source/WebKit/ModelProcess/cocoa/ModelProcessModelPlayerProxy.mm:
```diff
@@ -454,17 +455,28 @@ static CGFloat effectivePointsPerMeter(CALayer *caLayer)
+    m_interactionContainerEntity = interactionTarget;
+
+    // This also centers the interactionContainerEntity around the center of the model
+    m_stageModeInteractionDriver = adoptNS([WebKit::allocWKStageModeInteractionDriverInstance() initWithInteractionTarget:m_interactionContainerEntity.get() wkModel:m_modelRKEntity.get() container:m_containerEntity.get()]);
```
`m_containerEntity` is no longer in the entity hierarchy (I should have removed it as part of this: https://commits.webkit.org/289893@main - I'll look into cleaning that up in a separate PR or you can do it as part of this). But anyway, the model entity is a direct child of the layer's entity now.

Phipson replied:
Filed https://bugs.webkit.org/show_bug.cgi?id=288222 for follow-up

achan00 commented on Source/WebKit/ModelProcess/cocoa/ModelProcessModelPlayerProxy.mm:
```diff
@@ -675,17 +687,19 @@ static CGFloat effectivePointsPerMeter(CALayer *caLayer)
 void ModelProcessModelPlayerProxy::beginStageModeTransform(const WebCore::TransformationMatrix& transform)
 {
-    // FIXME: https://bugs.webkit.org/show_bug.cgi?id=287820
+    simd_float4x4 tf = simd_float4x4(transform);
```
Nit: should probably spell out `simdTransform` instead of `tf`

achan00 commented on Source/WebKit/ModelProcess/cocoa/ModelProcessModelPlayerProxy.mm:
```diff
@@ -675,17 +687,19 @@ static CGFloat effectivePointsPerMeter(CALayer *caLayer)
 void ModelProcessModelPlayerProxy::updateStageModeTransform(const WebCore::TransformationMatrix& transform)
 {
-    // FIXME: https://bugs.webkit.org/show_bug.cgi?id=287820
+    simd_float4x4 tf = simd_float4x4(transform);
```
Ditto.

achan00 commented on Source/WebKit/WebKitSwift/StageMode/WKStageMode.swift:
```diff
@@ -0,0 +1,109 @@
+internal import WebKitSwift
+internal import os
+internal import simd
+internal import Spatial
```
Nit: alphabetical ordering: should this go earlier with the other imports in uppercase?

achan00 commented on Source/WebKit/ModelProcess/StageMode/WKStageMode.h:
```diff
@@ -0,0 +1,50 @@
+};
+
+@interface WKStageModeInteractionDriver : NSObject
+- (instancetype)initWithInteractionTarget:(REEntityRef)target model:(WKSRKEntity * _Nullable)model container:(REEntityRef)container NS_SWIFT_NAME(init(with:model:container:));
```
I'd try to avoid passing core entity types in the WebKit APIs as `WKSRKEntity` is the WebKit wrapper for entities. There's an initializer in `WKSRKEntity` that takes in a `REEntityRef`.

Phipson replied:
I thought you said in an earlier PR to pass the WKSRKEntity, because I previously had 2 initializers, one that takes the `rootRKEntity` and another that takes the `rootEntity`.

achan00 commented on Source/WebKit/ModelProcess/StageMode/WKStageMode.h:
```diff
@@ -0,0 +1,50 @@
+/*
+ * Copyright (C) 2024 Apple Inc. All rights reserved.
```
Nit: 2025

achan00 commented on Source/WebKit/ModelProcess/StageMode/WKStageMode.swift:
```diff
@@ -0,0 +1,109 @@
+// Copyright (C) 2024 Apple Inc. All rights reserved.
```
Ditto.

achan00 commented on Source/WebKit/ModelProcess/StageMode/WKStageMode.h:
```diff
@@ -0,0 +1,50 @@
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#if defined(TARGET_OS_VISION) && TARGET_OS_VISION
```
Would `PLATFORM(VISION)` work?

achan00 replied:
Also, should this be guarded with `ENABLE(MODEL_PROCESS)`?

achan00 commented on Source/WebKit/ModelProcess/StageMode/WKStageMode.swift:
```diff
@@ -0,0 +1,109 @@
+internal import RealityKit
+internal import os
+internal import simd
+internal import Spatial
```
Nit: should `Spatial` go before `os`?

achan00 commented on Source/WebKit/ModelProcess/StageMode/WKStageMode.swift:
```diff
@@ -0,0 +1,109 @@
+@_spi(Private) @_spi(RealityKit) @_spi(CoreREAdditions) internal import RealityFoundation
+
+fileprivate extension Logger {
+    static let stageMode = Logger(subsystem: "com.apple.WebKit", category: "StageModeInteraction")
```
This is not used?

achan00 commented on Source/WebKit/WebKitSwift/WebKitSwift.h:
```diff
@@ -32,3 +32,4 @@
 #import <WebKitSwift/WKSLinearMediaTypes.h>
 #import <WebKitSwift/WKSPreviewWindowController.h>
 #import <WebKitSwift/WKSTextAnimationSourceDelegate.h>
+#import <WebKitSwift/WKStageMode.h>
```
Have you moved `WKStageMode.h` out of `WebKitSwift` already?

achan00 commented on Source/WebKit/ModelProcess/StageMode/WKStageMode.swift:
```diff
@@ -0,0 +1,109 @@
+        
+        let interactionContainer = Entity.__fromCore(__EntityRef.__fromCore(entity))
+        let modelEntity = Entity.__fromCore(__EntityRef.__fromCore(model.reEntity))
+        let containerEntity = Entity.__fromCore(__EntityRef.__fromCore(container))
```
If these entities are all passed in as `WKSRKEntity`, then we could expand `WKSRKEntity` to get visual bounds and set its parent so you can avoid converting between core and swift entity types.

Phipson replied:
Did this for the WKSRKEntity, but the container (i.e. the clientComponentEntity) still needs to be converted

achan00 commented on Source/WebKit/ModelProcess/cocoa/ModelProcessModelPlayerProxy.mm:
```diff
@@ -454,17 +455,28 @@ static CGFloat effectivePointsPerMeter(CALayer *caLayer)
+    m_interactionContainerEntity = interactionTarget;
+
+    // This also centers the interactionContainerEntity around the center of the model
+    m_stageModeInteractionDriver = adoptNS([WebKit::allocWKStageModeInteractionDriverInstance() initWithInteractionTarget:m_interactionContainerEntity.get() model:m_modelRKEntity.get() container:m_containerEntity.get()]);
```
`m_containerEntity` is no longer in the entity hierarchy (I should have removed it as part of this: https://commits.webkit.org/289893@main - I'll look into cleaning that up in a separate PR or you can do it as part of this). But anyway, the model entity is a direct child of the layer's entity now.

Phipson replied:
Filed https://bugs.webkit.org/show_bug.cgi?id=288222

achan00 commented on Source/WebKit/ModelProcess/StageMode/WKStageMode.swift:
```diff
@@ -0,0 +1,102 @@
+    
+    @objc override init() {
+        let interactionContainer = Entity()
+        let turntableInteractionContainer = Entity()
```
Can we just set to the data members directly?

achan00 commented on Source/WebKit/ModelProcess/StageMode/WKStageMode.h:
```diff
@@ -0,0 +1,50 @@
+};
+
+@interface WKStageModeInteractionDriver : NSObject
+- (instancetype)initWithModel:(WKSRKEntity * _Nullable)model container:(REEntityRef)container NS_SWIFT_NAME(init(with:container:));
```
Also, should we allow `model` to be nullable? Or it shouldn't be nullable to make it clear that `WKStageModeInteractionDriver` can only be created with a `model`?

achan00 commented on Source/WebKit/ModelProcess/StageMode/WKStageMode.h:
```diff
@@ -0,0 +1,50 @@
+
+NS_ASSUME_NONNULL_END
+
+#endif // defined(TARGET_OS_VISION) && TARGET_OS_VISION
```
Nit: the comment needs to be updated.

achan00 reviewed (commented):
Have a few comments but overall looks good. To @rr-codes for feedback in the Swift implementation setup.

emw-apple commented on Source/WebKit/Configurations/Base.xcconfig:
```diff
@@ -159,7 +159,14 @@ WK_ENABLE_WEB_EXTENSION_API[sdk=xr*2.1*] = NO;
+WK_MODEL_PROCESS_DEFINES = $(WK_MODEL_PROCESS_DEFINES_$(WK_ENABLE_MODEL_PROCESS))
+WK_MODEL_PROCESS_DEFINES_YES = WK_ENABLE_MODEL_PROCESS
+
+SWIFT_ACTIVE_COMPILATION_CONDITIONS = $(inherited) $(SWIFT_ACTIVE_COMPILATION_CONDITIONS_$(WK_ENABLE_SWIFTUI)) $(WK_MODEL_PROCESS_DEFINES)
```
nit: I would simplify this a bit so that it's similar to how WK_ENABLE_SWIFTUI works. You don't need to provide a declaration for the "NO" case. Typically, build settings are undefined when we want them to be in a default state.

Please also be sure to use semicolons, as that is our styling convention. I think you also typo'd the name of the SwiftUI setting, it needs to start with "SWIFT_ACTIVE_COMPILATION_CONDITIONS_SWIFTUI_".

```suggestion
WK_ENABLE_MODEL_PROCESS[sdk=xr*3*] = YES;

SWIFT_ACTIVE_COMPILATION_CONDITIONS = $(inherited) $(SWIFT_ACTIVE_COMPILATION_CONDITIONS_SWIFTUI_$(WK_ENABLE_SWIFTUI)) $(SWIFT_
ACTIVE_COMPILATION_CONDITONS_MODEL_PROCESS_$(WK_ENABLE_MODEL_PROCESS));
SWIFT_ACTIVE_COMPILATION_CONDITIONS_MODEL_PROCESS_YES = WK_ENABLE_MODEL_PROCESS;
```

achan00 replied:
Should we make sure `WK_ENABLE_MODEL_PROCESS`'s definition matches `ENABLE_MODEL_PROCESS`?

emw-apple reviewed (approved):
Build changes look fine, but please fix this issue before landing:

achan00 reviewed (approved):
I have a question about the `WK_ENABLE_MODEL_PROCESS` flag. But overall the PR looks good.

Phipson committed:
Create a driver to update StageMode Interactions based on Gesture Input
https://bugs.webkit.org/show_bug.cgi?id=287820
rdar://145006262

Reviewed by Wenson Hsieh, Richard Robinson, Elliott Williams, and Ada Chan.

This PR adds the StageModeInteractionDriver, which is a helper class object used in the ModelProcessModelPlayerProxy
to control how the model responds to the stagemode transform updates, based on the applied stagemode attribute. When
stagemode is set to 'none' we ignore all transforms passed to the player proxy. When stagemode is set to 'orbit' we
use the translation from the transform to rotate the model. To separate stagemode updates from Javascript updates,
we introduce an Interaction Container that wraps around the model entity, on which the stagemode updates will be applied.

The driver will be implemented in https://bugs.webkit.org/show_bug.cgi?id=287601.

* Source/WebKit/ModelProcess/cocoa/ModelProcessModelPlayerProxy.h:
* Source/WebKit/ModelProcess/cocoa/ModelProcessModelPlayerProxy.mm:
(WebKit::ModelProcessModelPlayerProxy::didFinishLoading):
(WebKit::ModelProcessModelPlayerProxy::beginStageModeTransform):
(WebKit::ModelProcessModelPlayerProxy::updateStageModeTransform):
(WebKit::ModelProcessModelPlayerProxy::endStageModeInteraction):
(WebKit::ModelProcessModelPlayerProxy::setStageMode):
(WebKit::ModelProcessModelPlayerProxy::applyStageModeOperationToDriver):
- Updates the callbacks to use the interaction driver
- Also adds helper function to update the driver's internal state based on the stagemode attribute

* Source/WebKit/UIProcess/Cocoa/WebKitSwiftSoftLink.h:
* Source/WebKit/UIProcess/Cocoa/WebKitSwiftSoftLink.mm:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebKitSwift/StageMode/WKStageMode.swift: Added.
(WKStageModeInteractionDriver.stageModeOperation):
(WKStageModeInteractionDriver.interactionDidBegin(_:)):
(WKStageModeInteractionDriver.interactionDidUpdate(_:)):
(WKStageModeInteractionDriver.interactionDidEnd):
(WKStageModeInteractionDriver.operationDidChange(_:)):
- Adds the necessary APIs to apply the transform to the interaction target.

* Source/WebKit/WebKitSwift/StageMode/WKStageModeBridging.h: Copied from Source/WebKit/WebKitSwift/WebKitSwift.h.
* Source/WebKit/WebKitSwift/WebKitSwift.h:

Canonical link: https://commits.webkit.org/291028@main
